### PR TITLE
Pipeline revng passes

### PR DESCRIPTION
**Pull request by drblallo**

### CODE REVIEW

File: lib/Pipeline/GenericLLVMPipe.cpp [hunk 1]
```diff
@@ -29,10 +29,49 @@ std::unique_ptr<LLVMPassWrapperBase> PureLLVMPassWrapper::clone() const {
...
+char AdvanceContract::ID = '_';
+
+static RP<AdvanceContract>
+  X("advance-contract", "Advances pipeline contracts", true, false);
+
+void GenericLLVMPipe::run(ExecutionContext &Ctx, LLVMContainer &Container) {
   llvm::legacy::PassManager Manager;
-  for (const auto &Element : Passes)
+  Manager.add(new LoadExecutionContextPass(&Ctx, Container.name()));
+  for (const auto &Element : Passes) {
```
Comment (aleclearmind) [inline lib/Pipeline/GenericLLVMPipe.cpp:69]: Using `auto` makes the code harder to understand

File: lib/Pipes/RevngPasses.cpp [hunk 2]
```diff
@@ -0,0 +1,49 @@
+//
```
Comment (aleclearmind) [inline lib/Pipes/RevngPasses.cpp:1]: Missing header for `.cpp` files.

File: lib/Pipeline/GenericLLVMPipe.cpp [hunk 3]
```diff
@@ -29,10 +29,49 @@ std::unique_ptr<LLVMPassWrapperBase> PureLLVMPassWrapper::clone() const {
   return std::make_unique<PureLLVMPassWrapper>(*this);
 }
 
-void GenericLLVMPipe::run(const ExecutionContext &, LLVMContainer &Container) {
+class AdvanceContract : public llvm::ModulePass {
```
Comment (aleclearmind) [inline lib/Pipeline/GenericLLVMPipe.cpp:32]: `MoveToNextContract`? `UpdateContract`? The contract itself does not move.

File: lib/Pipes/RevngPasses.cpp [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    Result = runOnTargets(Module, RequestedTargets);
+    Ctx->getContext().pushReadFields();
+    for (auto &Target : *RequestedTargets)
+      Ctx->commit(Target, Analysis.getWrittenContainerName());
+
+    Ctx->getContext().popReadFields();
+    Ctx->getContext().clearAndResume();
+
+  } else {
+    Result = runOnTargets(Module, nullptr);
```
Comment (aleclearmind) [inline lib/Pipes/RevngPasses.cpp:29]: What's the use case for *not* having a context?

File: lib/Pipes/RevngPasses.cpp [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+#include "revng/Pipes/RevngPasses.h"
+#include "revng/Pipes/TaggedFunctionKind.h"
+#include "revng/Support/FunctionTags.h"
+
+bool pipeline::RevngModulePass::runOnModule(llvm::Module &Module) {
+  auto &Analysis = getAnalysis<pipeline::LoadExecutionContextPass>();
+  ExecutionContext *Ctx = Analysis.get();
+  bool Result = false;
+  if (Ctx) {
+    auto name = Analysis.getWrittenContainerName();
```
Comment (aleclearmind) [inline lib/Pipes/RevngPasses.cpp:18]: `Name`

File: lib/Pipes/RevngPasses.cpp [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+  auto &Analysis = getAnalysis<pipeline::LoadExecutionContextPass>();
+  ExecutionContext *Ctx = Analysis.get();
+  auto &ModelWrapper = getAnalysis<LoadModelWrapperPass>().get();
+  bool Result = commonReads(Module, *ModelWrapper.getReadOnlyModel());
+
+  using Type = revng::kinds::TaggedFunctionKind;
+  auto ToIterOn = Type::getLLVMAndCommit(*Ctx,
+                                         Module,
+                                         Analysis.getWrittenContainerName());
+  for (auto Pair : ToIterOn) {
```
Comment (aleclearmind) [inline lib/Pipes/RevngPasses.cpp:44]: Use tuple unpacking

File: lib/Pipes/TaggedFunctionKind.cpp [hunk 7]
```diff
@@ -46,3 +46,50 @@ void TaggedFunctionKind::appendAllTargets(const pipeline::Context &Ctx,
...
   }
 }
+
+cppcoro::generator<std::pair<const model::Function *, llvm::Function *>>
+TaggedFunctionKind::getLLVMAndCommit(pipeline::ExecutionContext &Context,
+                                     llvm::Module &Module,
+                                     llvm::StringRef ContainerName) {
+  std::map<MetaAddress, llvm::Function *> AddressToFunction;
+
+  for (auto &LLVMFun : Module.functions()) {
```
Comment (aleclearmind) [inline lib/Pipes/TaggedFunctionKind.cpp:56]: 1. No abbreviations
2. `auto` abuse

File: lib/Pipes/TaggedFunctionKind.cpp [hunk 8]
```diff
@@ -46,3 +46,50 @@ void TaggedFunctionKind::appendAllTargets(const pipeline::Context &Ctx,
...
+    }
+
+    AddressToFunction.emplace(getMetaAddressOfIsolatedFunction(LLVMFun),
+                              &LLVMFun);
+  }
+  auto &Binary = getModelFromContext(Context);
+
+  for (const Target &Target :
+       Context.getCurrentRequestedTargets(ContainerName)) {
+    auto MetaAddres = MetaAddress::fromString(Target.getPathComponents()[0]);
```
Comment (aleclearmind) [inline lib/Pipes/TaggedFunctionKind.cpp:68]: `MetaAddress`, not `MetaAddres`.
This typo is present elsewhere too.

File: include/revng/Pipes/TaggedFunctionKind.h [hunk 9]
```diff
@@ -54,6 +55,16 @@ class TaggedFunctionKind : public pipeline::LLVMKind {
...
+  static cppcoro::generator<
+    std::pair<const model::Function *, llvm::Function *>>
+  getLLVMAndCommit(pipeline::ExecutionContext &Context,
+                   llvm::Module &Module,
+                   llvm::StringRef ContainerName);
+
+  static cppcoro::generator<const model::Function *>
+  getAndCommit(pipeline::ExecutionContext &Context,
+               const pipeline::ContainerBase &Container);
 };
```
Comment (aleclearmind) [inline include/revng/Pipes/TaggedFunctionKind.h:69]: I'd rename both of these into ``getFunctionsAndCommit`.
The arguments automatically disambiguate what's happening.

File: include/revng/Pipes/RevngPasses.h [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+class RevngModulePass : public llvm::ModulePass {
+  using llvm::ModulePass::ModulePass;
+
+  bool runOnModule(llvm::Module &Module) final;
+
+  virtual bool runOnTargets(llvm::Module &Module,
+                            const TargetsList *Targets) = 0;
+};
+
+class RevngFunctionPipe {
```
Comment (aleclearmind) [inline include/revng/Pipes/RevngPasses.h:23]: 1. Rename to `FunctionPassImpl`
2. Rename `RevngFunctionPass` to `FunctionPass`

**Review by aleclearmind — CHANGES_REQUESTED** [review]

File: lib/Pipes/RevngPasses.cpp (hunk 4 above)
Comment (drblallo) [reply lib/Pipes/RevngPasses.cpp:29]: forgot to remove the if from the version with opt support. Dropped now.

File: include/revng/FunctionIsolation/EnforceABI.h [hunk 11]
```diff
@@ -10,14 +10,25 @@
...
 public:
-  EnforceABI() : ModulePass(ID) {}
+  static char ID;
 
-  bool runOnModule(llvm::Module &M) override;
+public:
+  EnforceABI() : pipeline::FunctionPass<EnforceABIPassImpl>(ID) {}
 
   void getAnalysisUsage(llvm::AnalysisUsage &AU) const override;
 };
```
Comment (aleclearmind) [inline include/revng/FunctionIsolation/EnforceABI.h:34]: This has to go. It does nothing except inheriting from `pipeline::FunctionPass<EnforceABIPassImpl>`.
We already have to define the pipe and the `FunctionPassImpl`, not the `FunctionPass` too.
We can push `getAnalysisUsage` into `FunctionPassImpl` too, if that's the reason.

File: include/revng/FunctionIsolation/InvokeIsolatedFunctions.h [hunk 12]
```diff
@@ -11,18 +11,19 @@
...
-  InvokeIsolatedFunctionsPass() : ModulePass(ID) {}
+  InvokeIsolatedFunctionsPass() : pipeline::ModulePass(ID) {}
 
-  bool runOnModule(llvm::Module &M) override;
+  bool run(llvm::Module &M, const pipeline::TargetsList &Targets) override;
 
   void getAnalysisUsage(llvm::AnalysisUsage &AU) const override {
     AU.addRequired<GeneratedCodeBasicInfoWrapperPass>();
     AU.addRequired<LoadModelWrapperPass>();
+    AU.addRequired<pipeline::LoadExecutionContextPass>();
```
Comment (aleclearmind) [inline include/revng/FunctionIsolation/InvokeIsolatedFunctions.h:26]: Why do we still need `LoadModelWrapperPass`?
We can get the model from the `ExecutionContext`, can't we?

File: include/revng/Model/IRHelpers.h [hunk 13]
```diff
@@ -22,9 +22,8 @@ inline model::Function *llvmToModelFunction(model::Binary &Binary,
   auto MaybeMetaAddress = getMetaAddressMetadata(&F, FunctionEntryMDNName);
   if (MaybeMetaAddress == MetaAddress::invalid())
     return nullptr;
-  if (auto It = Binary.Functions().find(MaybeMetaAddress);
-      It != Binary.Functions().end())
-    return &*It;
+  if (auto It = Binary.Functions().tryGet(MaybeMetaAddress); It != nullptr)
+    return It;
```
Comment (aleclearmind) [inline include/revng/Model/IRHelpers.h:26]: This change should be a separate commit.

File: include/revng/ADT/TrackingContainer.h [hunk 14]
```diff
@@ -266,6 +266,18 @@ class TrackingContainer {
...
+    if (Iter == Content.end()) {
+      if (TrackingIsActive)
+        NonExisting.back().insert(Key);
+      return nullptr;
+    }
+
+    markExistingKey(Key);
+    return &*Iter;
+  }
+
```
Comment (aleclearmind) [inline include/revng/ADT/TrackingContainer.h:280]: This should be a separate commit.

File: include/revng/TupleTree/Visits.h [hunk 15]
```diff
@@ -271,6 +272,10 @@ bool callOnPathSteps(Visitor &V,
   return tupletree::detail::callOnPathStepsTuple(V, Path, M, FullPath);
 }
 
+template<typename T>
+concept IsTrackingContainer = StrictSpecializationOf<T,
+                                                     revng::TrackingContainer>;
```
Comment (aleclearmind) [inline include/revng/TupleTree/Visits.h:277]: I'd check if it has `tryGet`.

File: lib/EarlyFunctionAnalysis/CollectFunctionsFromUnusedAddressesPass.cpp [hunk 16]
```diff
@@ -68,7 +68,7 @@ class CFFUAImpl {
         continue;
 
       MetaAddress Entry = getBasicBlockAddress(getJumpTargetBlock(&BB));
-      if (Binary.Functions().find(Entry) != Binary.Functions().end())
+      if (Binary.Functions().tryGet(Entry) != nullptr)
```
Comment (aleclearmind) [inline lib/EarlyFunctionAnalysis/CollectFunctionsFromUnusedAddressesPass.cpp:71]: Switching to `tryGet`should get its own commit.

File: lib/FunctionIsolation/IsolateFunctions.cpp [hunk 17]
```diff
@@ -675,7 +675,7 @@ void IsolateFunctionsImpl::run() {
   FunctionTags::IsolatedRoot.addTo(RootFunction);
 }
 
-bool IF::runOnModule(Module &TheModule) {
+bool IF::run(Module &TheModule, const pipeline::TargetsList &Targets) {
```
Comment (aleclearmind) [inline lib/FunctionIsolation/IsolateFunctions.cpp:678]: We should use the targets. At least leave a TODO.

File: lib/FunctionIsolation/PromoteCSVs.cpp [hunk 18]
```diff
@@ -572,26 +576,40 @@ void PromoteCSVs::wrapCallsToHelpers(Function *F) {
...
-bool PromoteCSVsPass::runOnModule(Module &M) {
+bool PromoteCSVsPassImpl::prologue(llvm::Module &Module,
+                                   const model::Binary &Binary) {
+
   auto &GCBI = getAnalysis<GeneratedCodeBasicInfoWrapperPass>().getGCBI();
   auto &ModelWrapper = getAnalysis<LoadModelWrapperPass>().get();
-  const model::Binary &Binary = *ModelWrapper.getReadOnlyModel();
-  PromoteCSVs HW(&M, GCBI, Binary);
-  HW.run();
+  HW = new PromoteCSVs(&Module, GCBI, Binary);
```
Comment (aleclearmind) [inline lib/FunctionIsolation/PromoteCSVs.cpp:603]: Absolutely not.

File: lib/FunctionIsolation/PromoteCSVs.cpp [hunk 19]
```diff
@@ -572,26 +576,40 @@ void PromoteCSVs::wrapCallsToHelpers(Function *F) {
...
+    // (Re-)promote CSVs
+    promoteCSVs(&F);
+  }
+}
+
+void PromoteCSVs::run() {
+  for (Function &F : FunctionTags::ABIEnforced.functions(M)) {
+    runOnFunction(F);
   }
 }
```
Comment (aleclearmind) [inline lib/FunctionIsolation/PromoteCSVs.cpp:596]: Drop?

File: lib/Pipeline/Target.cpp [hunk 20]
```diff
@@ -103,25 +103,25 @@ llvm::Error pipeline::parseTarget(const Context &Ctx,
...
+  Name.split(Path, '/');
 
-  auto It = llvm::find_if(Dict,
-                          [&Parts](Kind &K) { return Parts[1] == K.name(); });
+  auto It = llvm::find_if(Dict, [&](Kind &K) { return KindName == K.name(); });
   if (It == Dict.end())
     return llvm::createStringError(llvm::inconvertibleErrorCode(),
                                    "No known Kind '%s' in dictionary",
-                                   Parts[1].str().c_str());
+                                   KindName.str().c_str());
```
Comment (aleclearmind) [inline lib/Pipeline/Target.cpp:124]: This probably deserves its own commit.

**Review by aleclearmind — CHANGES_REQUESTED** [review]

File: lib/FunctionIsolation/PromoteCSVs.cpp (hunk 18 above)
Comment (drblallo) [reply lib/FunctionIsolation/PromoteCSVs.cpp:603]: i can remove the new invocation here, but it would imply that PromoteCSVs is moved to a header

File: include/revng/FunctionIsolation/InvokeIsolatedFunctions.h (hunk 12 above)
Comment (drblallo) [reply include/revng/FunctionIsolation/InvokeIsolatedFunctions.h:26]: yes, but we need to schedule LOadModelWrapperPAss anyway so people can do both. Do we wish to have a policy about this? in principle we could even entirely drop LoadModelWrapperPass

File: include/revng/TupleTree/Visits.h (hunk 15 above)
Comment (drblallo) [reply include/revng/TupleTree/Visits.h:277]: i considered it, but i think this is more correct, bucause the issue arises not from people using vectors that do not have try get, but from the fact that one may generate a model like from the yaml without instrumentation, so here we actually want to differentiate those to cases.

File: lib/FunctionIsolation/IsolateFunctions.cpp [hunk 21]
```diff
@@ -540,7 +544,10 @@ void IsolateFunctionsImpl::run() {
   //
   // Precreate the isolated functions
   //
-  for (const model::Function &Function : Binary.Functions()) {
+  for (const pipeline::Target &Target :
+       Targets->filterByKind(revng::kinds::Isolated)) {
+    auto MetaAddress = MetaAddress::fromString(Target.getPathComponents()[0]);
+    const auto &Function = Binary.Functions().at(MetaAddress);
```
Comment (aleclearmind) [inline lib/FunctionIsolation/IsolateFunctions.cpp:550]: Revert. We'll take care of this later.

**Review by aleclearmind — CHANGES_REQUESTED** [review]

File: include/revng/FunctionIsolation/PromoteCSVs.h [hunk 22]
```diff
@@ -8,17 +8,29 @@
 
 #include "revng/BasicAnalyses/GeneratedCodeBasicInfo.h"
 
-class PromoteCSVsPass : public llvm::ModulePass {
+class PromoteCSVs;
+class PromoteCSVsPassImpl : public pipeline::FunctionPassImpl {
 public:
-  static char ID;
+  PromoteCSVs *HW;
```
[the remark this answers is not in the document]
Comment (drblallo) [reply include/revng/FunctionIsolation/PromoteCSVs.h]: resolved by pushing promotecsv in cpp file

File: include/revng/FunctionIsolation/InvokeIsolatedFunctions.h (hunk 12 above)
Comment (drblallo) [reply include/revng/FunctionIsolation/InvokeIsolatedFunctions.h:26]: i looked into this, but the change was sprawling and it still requires to make sure that revng opt is never used. I recommend to do it after after that issue is fixed

**Comment by aleclearmind** [comment]
Closing this due to the heavy reworking I'm doing on this branch.
